### PR TITLE
Add wandb support

### PR DESCRIPTION
**Pull request by gabrieltseng**
- [x] Run on the train and val parquet files
- [x] Log all results to wandb 
- [x] (initial) speed up of the dataset
- [x] WorldCereals eval task  

Tackles #6 , and to a lesser extent #2 (epochs are about twice as fast, but GPU utilization remains low, and training time per-instance is still very slow).

### CODE REVIEW

File: src/eval.py [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+                )
+                encoding_list.append(encodings)
+        encodings_np = np.concatenate(encoding_list)
+        targets = np.concatenate(target_list)
+        if len(targets.shape) == 2 and targets.shape[1] == 1:
+            targets = targets.ravel()
+
+        fit_models = []
+        model_dict = {
+            "Regression": LogisticRegression(
```
Comment (kvantricht) [inline src/eval.py:71]: for my understanding, what's the purpose of the logistic regression task?

File: src/dataset.py [hunk 2]
```diff
@@ -1,184 +1,123 @@
...
+from .dataops import BANDS, S1_S2_ERA5_SRTM, DynamicWorld2020_2021
 from .masking import MaskedExample, MaskParamsNoDw
-from .utils import construct_single_presto_input
 
 
-class WorldCerealDataset(Dataset):
-    def __init__(self, dataframe: pd.DataFrame, mask_params: Optional[MaskParamsNoDw] = None):
+class WorldCerealBase(Dataset):
+    _NODATAVALUE = 65535
+    NUM_TIMESTEPS = 12
```
[the remark this answers is not in the document]
Comment (gabrieltseng) [reply src/dataset.py:14]: Yes, I agree (about masking). I think this would not impact the pre-training portion of the code, so we can probably tackle that with a mask in the labelled dataset.

File: src/dataset.py [hunk 3]
```diff
@@ -1,184 +1,123 @@
...
-        return self.convert_inputs(row)
-
-    def convert_inputs(self, row):
-
-        # Latitude/Longitude
-        latlon = torch.tensor([row.lat, row.lon]).float()
-
-        # Month
+    @classmethod
+    def row_to_arrays(cls, row: pd.Series) -> Tuple[np.ndarray, np.ndarray, float, int]:
```
[the remark this answers is not in the document]
Comment (gabrieltseng) [reply src/dataset.py:23]: Yes - I think one of the obvious things to tackle as part of #2 (cc @rubencart ) is to at least cache the arrays so they don't need to be recomputed each epoch.

File: src/dataset.py [hunk 4]
```diff
@@ -1,184 +1,123 @@
...
+                values[idx_valid] = values[idx_valid] / (100 * 1000.0)
+            elif presto_val == "temperature_2m":
+                # remove scaling
+                values[idx_valid] = values[idx_valid] / 100
+            eo_data[:, BANDS.index(presto_val)] = values
+        static_band_mapping = {"DEM-alt-20m": "elevation", "DEM-slo-20m": "slope"}
+        for df_val, presto_val in static_band_mapping.items():
+            eo_data[:, BANDS.index(presto_val)] = row[df_val]
+
+        return eo_data, latlon, month, row["LANDCOVER_LABEL"] == 11
```
[the remark this answers is not in the document]
Comment (gabrieltseng) [reply src/dataset.py:62]: Yes - it occurs right when the dataset is initialized ([lines 106 to 109](https://github.com/WorldCereal/presto-worldcereal/blob/e5125326a3abc7b36f954d66647ee1e896feef70/src/dataset.py#L107)).

File: src/dataset.py [hunk 5]
```diff
@@ -1,184 +1,123 @@
...
 
-            single_sample[b] = values
+    def __getitem__(self, idx):
+        raise NotImplementedError
 
-        s1_data = torch.from_numpy(single_sample.values).float()
+    @classmethod
+    def normalize_and_mask(cls, eo: np.ndarray):
```
[the remark this answers is not in the document]
Comment (gabrieltseng) [reply src/dataset.py:68]: Right now, we set them to `0`. 

In this part of the function, they remain `_NODATAVALUE` - then, in `normalize_and_mask` that value gets set to `0`. `normalize_and_mask` gets called in the classes which implement `getitem` (`WorldCerealMaskedDataset` and `WorldCerealLabelledDataset`).

I did it this way because I want us to be able to swap this out for a "True" mask next.

File: src/eval.py (hunk 1 above)
Comment (gabrieltseng) [reply src/eval.py:71]: This is just legacy code - its implemented for free in the [main presto codebase](https://github.com/nasaharvest/presto/blob/8d8563ab680160edfded1dbf153c0f8b0b866c93/presto/eval/eval.py#L160), and I thought we might want to run this as a quick approximation of finetuning Presto. This is why I left it in rather than deleting it.

File: train.py [hunk 6]
```diff
@@ -90,15 +171,18 @@
...
 
+if wandb_enabled:
+    wandb.config.update(training_config)
+
 lowest_validation_loss = None
 best_val_epoch = 0
 training_step = 0
 num_validations = 0
-dataloader_length = df.shape[0]
+dataloader_length = train_df.shape[0]
```
Comment (rubencart) [inline train.py:185]: Shouldn't this be in batches? Cfr https://github.com/WorldCereal/presto-worldcereal/blob/304b2fb33311e50085e2143af9e659322cdd6abd/train.py#L194-L202

File: src/dataset.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (kvantricht) [reply src/dataset.py:68]: Ah ok makes sense!

File: train.py (hunk 6 above)
Comment (kvantricht) [reply train.py:185]: I think indeed I also computed the `dataloader_length` by dividing the the dataframe length by the batch size.

Comment (gabrieltseng) [reply train.py:185]: doh! nice catch. Fixed in https://github.com/WorldCereal/presto-worldcereal/pull/8

File: train.py [hunk 7]
```diff
@@ -1,17 +1,21 @@
...
 from pathlib import Path
+from typing import Tuple, cast
 
 import pandas as pd
 import torch
 import torch.nn as nn
 from torch import optim
 from torch.utils.data import DataLoader
 from tqdm import tqdm
+from wandb.sdk.wandb_run import Run
```
Comment (rubencart) [inline train.py:14]: This fails if you don't have wandb, so no point in doing
```python
if wandb_enabled:
    import wandb
```
then :-).
